### PR TITLE
[Core] Refactor CommandLine arguments

### PR DESCRIPTION
**Pull request by ltpquang**
### Problem

I'm about to call Cucumber `main` function from my own app

```
public static void main(String[] args) {
        io.cucumber.core.cli.Main.main(args);
}
```

For better management, I would standardize my app's arguments, before passing to Cucumber.

The problem is Cucumber's arguments are now hardcoded, which make future updates maybe break my app.

### Solution

Define all arguments as static string variables. This makes arguments more consistent, and cleaner codebase.

### CODE REVIEW

File: core/src/main/java/io/cucumber/core/options/CommandlineOptions.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package io.cucumber.core.options;
+
+public class CommandlineOptions {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/options/CommandlineOptions.java:3]: - This class should be final. 
- Moved to `io.cucumber.core.cli`
- Annotated with `@Api` (same as the Main class)
- Add a private constructor to prevent instantiation

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
This is useful. Since this class is to be used in combination with the main class it would make sense to put it in the same package. It is essentially an interface for public consumption..

File: core/src/main/java/io/cucumber/core/options/CommandlineOptions.java (hunk 1 above)
Comment (mpkorstanje) [reply core/src/main/java/io/cucumber/core/options/CommandlineOptions.java:3]: And don't forget to add a line to the `CHANGELOG`

Comment (ltpquang) [reply core/src/main/java/io/cucumber/core/options/CommandlineOptions.java]: @mpkorstanje good point on private constructor 👍 I've updated the PR as your suggestions, please help me review them. Thanks!

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Changed
 
+ * [Core] Refactor CommandLine arguments ([2130](https://github.com/cucumber/cucumber-jvm/pull/2130) Quang Lê)
```
Comment (mpkorstanje) [inline CHANGELOG.md:14]: The change log is read from the perspective of a user of the library. From that perspective you did not refactor any thing. Rather you added constants for use with the CLI.

File: core/src/main/java/io/cucumber/core/cli/CLIOptions.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
+package io.cucumber.core.cli;
+
+import org.apiguardian.api.API;
+
+@API(status = API.Status.STABLE)
+public final class CLIOptions {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/cli/CLIOptions.java:6]: CommandLineOptions was a good name. Why did you change it?

Comment (mpkorstanje) [reply core/src/main/java/io/cucumber/core/cli/CLIOptions.java:6]: And don't forget to add some java doc. Gotta explain to people what this is for. And what to use it in combination with.

File: core/src/main/java/io/cucumber/core/options/CommandlineOptionsParser.java [hunk 4]
```diff
@@ -65,66 +66,70 @@ private RuntimeOptionsBuilder parse(List<String> args) {
         while (!args.isEmpty()) {
             String arg = args.remove(0).trim();
 
-            if (arg.equals("--help") || arg.equals("-h")) {
+            if (arg.equals(CLIOptions.HELP) || arg.equals(CLIOptions.HELP_SHORT)) {
                 printUsage();
                 exitCode = 0;
                 return parsedOptions;
-            } else if (arg.equals("--version") || arg.equals("-v")) {
+            } else if (arg.equals(CLIOptions.VERSION) || arg.equals(CLIOptions.VERSION_SHORT)) {
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/options/CommandlineOptionsParser.java:73]: For readability static imports would be useful

**Review by mpkorstanje — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (ltpquang) [reply CHANGELOG.md]: yeah I've just updated it for easier to understand 👍

File: core/src/main/java/io/cucumber/core/options/CommandlineOptionsParser.java (hunk 4 above)
Comment (ltpquang) [reply core/src/main/java/io/cucumber/core/options/CommandlineOptionsParser.java]: static import is good, I've just updated that

File: core/src/main/java/io/cucumber/core/cli/CLIOptions.java (hunk 3 above)
Comment (ltpquang) [reply core/src/main/java/io/cucumber/core/cli/CLIOptions.java]: I renamed because I didn't use static import at first, making the code a little bit clumsy. Since I'd used static import, I reverted it to the old name. And included some javadoc too. 😄

File: core/src/main/java/io/cucumber/core/cli/CommandlineOptions.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Contains all available command line options for {@link io.cucumber.core.cli.Main}
+ * <p>
+ * After being passed to {@link io.cucumber.core.cli.Main#main} function, these options will be
+ * parsed by {@link io.cucumber.core.options.CommandlineOptionsParser} to provide running options
+ * to CLI.
+ * <p>
+ * All the options are defined as static string variables to allow other programs to call
+ * {@link io.cucumber.core.cli.Main#main} function in a more consistent way.
+ * E.g. {@code io.cucumber.core.cli.Main.main(NAME, "TestName", THREADS, 2)}
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/cli/CommandlineOptions.java:14]: Note that Main.main will call ` System.exit(exitStatus)`. This may not be desirable for most users. Would it be worth adding a `Main.run(String... argv)`?

Comment (mpkorstanje) [reply core/src/main/java/io/cucumber/core/cli/CommandlineOptions.java:14]: Also `"2"`

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
Almost there!

**Comment by mpkorstanje** [comment]
I pushed some changes. Let me know what you you think.

**Comment by ltpquang** [comment]
> I pushed some changes. Let me know what you you think.

Thanks @mpkorstanje , everything seems to be great now. But for later PRs, what changes will be categorized as `Added`, and what will be categorized as `Changed`, in the CHANGELOG? Since my PR _changes_ some behavior, by _adding_ some variables, this makes me a little bit confused.

**Comment by mpkorstanje** [comment]
We're using Keep a Changelog as a guide: https://keepachangelog.com/en/1.0.0/

> Types of changes
> 
>     Added for new features.
>     Changed for changes in existing functionality.
>     Deprecated for soon-to-be removed features.
>     Removed for now removed features.
>     Fixed for any bug fixes.
>     Security in case of vulnerabilities.

Since this wasn't existing behaviour, changed does not apply.

**Comment by ltpquang** [comment]
> We're using Keep a Changelog as a guide: https://keepachangelog.com/en/1.0.0/
> 
> > Types of changes
> > ```
> > Added for new features.
> > Changed for changes in existing functionality.
> > Deprecated for soon-to-be removed features.
> > Removed for now removed features.
> > Fixed for any bug fixes.
> > Security in case of vulnerabilities.
> > ```
> 
> Since this wasn't existing behaviour, changed does not apply.

@mpkorstanje Got it. Everything is clear by now, please continue PR process as I don't have any further comments. Thank you!

**Comment by aslakhellesoy** [comment]
Hi @ltpquang,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber
